controller/test/frame_detector/frame_detector_test.cpp

```diff
@@ -0,0 +1,198 @@
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::DMA);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::OVERRUN);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::SERIAL_FRAMING);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::UNKNOWN);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+}
+
+vector<string> fakeRx(string frame) {
+  vector<string> ret;
+  FrameDetectorTest frame_detector(rx_buf);
+  if (!frame_detector.Begin()) {
+    return ret;
+  }
+
+  for (auto &c : frame) {
+    rx_buf.PutByte(c);
+    if (frame_detector.is_frame_available()) {
+      string s;
+      auto length = static_cast<size_t>(frame_detector.get_frame_length());
+      s.assign(reinterpret_cast<char *>(frame_detector.TakeFrame()), length);
+      ret.push_back(s);
+    }
+  }
+
+  return ret;
+}
+
+TEST(FrameDetector, FuzzMoreMarkers) {
+  EXPECT_THAT(fakeRx(".aaa..."), testing::ElementsAre("aaa"));
+  EXPECT_THAT(fakeRx("..aaa...."), testing::ElementsAre("aaa"));
+  EXPECT_THAT(fakeRx("...aaa.."), testing::ElementsAre("aaa"));
+  EXPECT_THAT(fakeRx("....aaa."), testing::ElementsAre("aaa"));
+}
+
+TEST(FrameDetector, FramesTooLong) {
+  EXPECT_THAT(fakeRx(".aaaaaaaaaaaaaaaaaaaa..aaa."),
+              testing::ElementsAre("aaa"));
+  EXPECT_THAT(fakeRx(".aaaaaaaaaaaaaaaaaaa..aaa."),
+              testing::ElementsAre("aaa"));
+  EXPECT_THAT(fakeRx("....................aaa."), testing::ElementsAre("aaa"));
+  EXPECT_THAT(fakeRx("...................aaa."), testing::ElementsAre("aaa"));
+}
+
+TEST(FrameDetector, FuzzyInputs) {
+  EXPECT_THAT(fakeRx(".aaa..bbb..a."), testing::ElementsAre("aaa", "bbb", "a"));
+  EXPECT_THAT(fakeRx("aaa.bbb.a."), testing::ElementsAre());
+  EXPECT_THAT(fakeRx("aaa.....bbb.a."), testing::ElementsAre("bbb"));
+  EXPECT_THAT(fakeRx(".aaa....bbb.a."), testing::ElementsAre("aaa", "bbb"));
+  EXPECT_THAT(fakeRx("aa.bbb.b...bbbb..s.ss.s...ssss..ss.ssss..aaaaa.sa.aaaa."),
+              testing::ElementsAre("bbbb", "s", "ssss", "ss", "aaaaa"));
+}
```

Comment:
These tests look awesome! I think we need just one more test verifying what I think is the most important property of this protocol - being able to recover from any errors. If we can do that, than we're safe, cause corrupt packets will get caught by CRC.

Take an example string S containing some valid packets - say, "..aaa..bbb..ccc.".
Now we want to verify that, no matter what good or bad stuff we prepend, the result **always** will still contain aaa, bbb, ccc.

We can do that by prepending several (say, up to 10) random events, where each event is either PutByte(marker), or PutByte(non-marker), or an error - i.e. 3 different possible events, 3^10 = 59049 possible prefixes. Also set LEN to a small value (say, 4) so that sometimes several non-markers in a row cover the LEN - 1, LEN - 2, LEN cases. For each of these 59049 prefixes, verify EXPECT_THAT(fakeRx(prefix + S), testing::Contains("aaa", "bbb", "ccc")).

This seems like it should be very simple to implement, but I'm fine if it's in a separate PR.

Reply:
DONE